### PR TITLE
fix(bigquery): Job Library fixes for InsertJob api

### PR DESCRIPTION
**Pull request by jsrinnn**
This PR includes the following fixes for the core Job library, found during benchmark testing of the InsertJob api

- Fixed 64-bit parsing of the "maximum_bytes_billed" field
- Added support for removing keys from the Json payload that are not relevant for the api. 
  This is especially needed for insertjob api as it errors out if json fields with empty values are supplied. 
  For read-only apis we do want the json payload to return all fields.
- Modified validation logic in `BuildRestRequest` method for InsertJobRquest. The assumptions before were not valid. 
- Modified InsertJob() rest stub to filter json payload to only keys that are needed for that api.
- Corresponding unit test changes for all the above changes.

The above has been tested end to end with a benchmark program. I'll send the benchmark program changes after the above is merged.

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12746)

### CODE REVIEW

File: google/cloud/bigquery/v2/minimal/internal/json_utils.cc [hunk 1]
```diff
@@ -85,6 +85,32 @@ void ToJson(std::chrono::system_clock::time_point const& field,
...
 }
 
+nlohmann::json RemoveJsonKeysAndEmptyFields(std::string const& json_payload,
+                                            std::vector<std::string> keys) {
+  nlohmann::json::parser_callback_t remove_empty_call_back =
+      [&keys](int depth, nlohmann::json::parse_event_t event,
+              nlohmann::json& parsed) {
+        // Have this redundant check due to compile error that
+        // depth parameter is not being used.
+        if (depth < 0) return false;
```
Comment (dbolduc) [inline google/cloud/bigquery/v2/minimal/internal/json_utils.cc:95]: nit: Can you say `int /* depth */, ...` in the signature?

If not, I'd change this line to `(void)depth;`

File: google/cloud/bigquery/v2/minimal/internal/job_rest_stub.cc [hunk 2]
```diff
@@ -64,11 +64,14 @@ StatusOr<InsertJobResponse> DefaultBigQueryJobRestStub::InsertJob(
   nlohmann::json json_payload;
   to_json(json_payload, request.job());
 
+  auto filtered_json = RemoveJsonKeysAndEmptyFields(json_payload.dump(),
+                                                    request.json_filter_keys());
```
Comment (dbolduc) [inline google/cloud/bigquery/v2/minimal/internal/job_rest_stub.cc:68]: for the record, I don't think we ever `request.set_json_filter_keys(...)` so we are not removing any keys in this PR.

File: google/cloud/bigquery/v2/minimal/internal/json_utils.cc [hunk 3]
```diff
@@ -85,6 +85,32 @@ void ToJson(std::chrono::system_clock::time_point const& field,
   j[name] = std::to_string(m);
 }
 
+nlohmann::json RemoveJsonKeysAndEmptyFields(std::string const& json_payload,
+                                            std::vector<std::string> keys) {
```
Comment (dbolduc) [inline google/cloud/bigquery/v2/minimal/internal/json_utils.cc:89]: Hm so we are passing these keys in by value, but using them as a reference.

The time complexity is O(N*M) where N is number of keys, and M is number of fields.

I do not know how many keys you plan to skip (and whether they show up multiple times or not).

If the keys only show up once and you know their order, you could use something like a stack, where you discard the top element once you have found it. That would only be O(M).

File: google/cloud/bigquery/v2/minimal/internal/json_utils.cc [hunk 4]
```diff
@@ -85,6 +85,32 @@ void ToJson(std::chrono::system_clock::time_point const& field,
...
+  nlohmann::json::parser_callback_t remove_empty_call_back =
+      [&keys](int depth, nlohmann::json::parse_event_t event,
+              nlohmann::json& parsed) {
+        // Have this redundant check due to compile error that
+        // depth parameter is not being used.
+        if (depth < 0) return false;
+
+        if (event == nlohmann::json::parse_event_t::key) {
+          auto const key_found = std::any_of(
```
Comment (dbolduc) [inline google/cloud/bigquery/v2/minimal/internal/json_utils.cc:98]: This loop makes me nervous. See above.

File: google/cloud/bigquery/v2/minimal/internal/json_utils.cc [hunk 5]
```diff
@@ -85,6 +85,32 @@ void ToJson(std::chrono::system_clock::time_point const& field,
...
+              keys.begin(), keys.end(),
+              [&parsed](std::string const& key) { return (parsed == key); });
```
Comment (dbolduc) [inline google/cloud/bigquery/v2/minimal/internal/json_utils.cc:100]: nit: remove parentheses around the comparison

File: google/cloud/bigquery/v2/minimal/internal/json_utils.cc (hunk 4 above)
Comment (dbolduc) [reply google/cloud/bigquery/v2/minimal/internal/json_utils.cc:98]: Also I might rename `key_found` to something like `skip_key`

File: google/cloud/bigquery/v2/minimal/internal/json_utils.cc (hunk 3 above)
Comment (dbolduc) [reply google/cloud/bigquery/v2/minimal/internal/json_utils.cc:89]: Is N like... 2 or 3?

**Review by dbolduc — COMMENTED** [review]
It is unclear to me how we plan to use the filter keys. If the filter key business was in a separate PR, I could approve the other changes. :shrug:

File: google/cloud/bigquery/v2/minimal/internal/json_utils.cc [hunk 6]
```diff
@@ -85,6 +85,32 @@ void ToJson(std::chrono::system_clock::time_point const& field,
...
+          return !key_found;
+        }
+        if (event == nlohmann::json::parse_event_t::object_end) {
+          if (parsed == nullptr || parsed.empty()) return false;
+        }
+        if (event == nlohmann::json::parse_event_t::array_end) {
+          if (parsed.empty()) return false;
+        }
```
Comment (devbww) [inline google/cloud/bigquery/v2/minimal/internal/json_utils.cc:108]: It would be clearer to always return from each event-type block, rather than allowing any fall-through.

So, in the `object_end` case, for example, ...
```
    return parsed != nullptr && !parsed.empty();
```

**Review by jsrinnn — COMMENTED** [review]
Yes, the usage is in the benchmark PR which I'll be sending next. 

I was given the feedback to separate out the benchmark changes and core library changes. Hence two separate PRs.

Addressed all feedback. PTAL.

Thank you Darren and Bradley!

> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12746#-:-NfR47kx2s3dCdtTGkes:bband4t)* status: 0 of 10 files reviewed, 4 unresolved discussions (waiting on @dbolduc and @devbww)

___
*[`google/cloud/bigquery/v2/minimal/internal/job_rest_stub.cc` line 68 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12746#gh-1339469216:-NfQyBYB81MHRprFiGaR:b-dfnz2h) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/f050a835a3c70476b51b990846aa8828a7f16c69/google/cloud/bigquery/v2/minimal/internal/job_rest_stub.cc#L68)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

for the record, I don't think we ever `request.set_json_filter_keys(...)` so we are not removing any keys in this PR.
</blockquote></details>

That is being done in the benchmark tests. I am separating out the benchmark and the core library changes as was requested in the previous PR. I'll send the benchmark PR after this one.
___
*[`google/cloud/bigquery/v2/minimal/internal/json_utils.cc` line 89 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12746#gh-1339471504:-NfQymY3EA7yBJW5WNN8:b-o8m731) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/f050a835a3c70476b51b990846aa8828a7f16c69/google/cloud/bigquery/v2/minimal/internal/json_utils.cc#L89)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

Is N like... 2 or 3?
</blockquote></details>

N is less than 10 and is usually not repeated. I do not know their order. 

I have modified the signature to not pass by value. 
___
*[`google/cloud/bigquery/v2/minimal/internal/json_utils.cc` line 95 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12746#gh-1339462267:-NfR0B2Q0rDVbGtbUbeW:bseusp8) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/f050a835a3c70476b51b990846aa8828a7f16c69/google/cloud/bigquery/v2/minimal/internal/json_utils.cc#L95)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

nit: Can you say `int /* depth */, ...` in the signature?

If not, I'd change this line to `(void)depth;`
</blockquote></details>

Done

Ah...nice. Thank you. I wasn't aware of the `void` syntax
___
*[`google/cloud/bigquery/v2/minimal/internal/json_utils.cc` line 98 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12746#gh-1339472042:-NfR0RiY8vyNHyJro_Wt:bmbv2xi) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/f050a835a3c70476b51b990846aa8828a7f16c69/google/cloud/bigquery/v2/minimal/internal/json_utils.cc#L98)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

Also I might rename `key_found` to something like `skip_key`
</blockquote></details>

The benchmark testing results were comparable to the current version (of the driver) with the same input parameters. But I'll kepp this in mind in case we need to optimize.

`key_found` is the output of `std::any_of`  which according to the predicate indicates whether key was found or not.  

If the key is found we want to return false (i.e skip it)

 I can add a new variable here called `skip_key = !key_found` and return that so the semantics are clear that we are skipping the key incase key is found. The act of skipping the key involves returning false.

If you have a better alternative keeping the above semantics do let me know.
___
*[`google/cloud/bigquery/v2/minimal/internal/json_utils.cc` line 100 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12746#gh-1339472553:-NfR37BZ6wm8I6ggYBVL:bb6un1j) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/f050a835a3c70476b51b990846aa8828a7f16c69/google/cloud/bigquery/v2/minimal/internal/json_utils.cc#L100)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

nit: remove parentheses around the comparison
</blockquote></details>

Done
___
*[`google/cloud/bigquery/v2/minimal/internal/json_utils.cc` line 108 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12746#gh-1339491465:-NfR3l8N0sGPDrCdXw-w:b-lgb6m4) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/f050a835a3c70476b51b990846aa8828a7f16c69/google/cloud/bigquery/v2/minimal/internal/json_utils.cc#L108)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

It would be clearer to always return from each event-type block, rather than allowing any fall-through.

So, in the `object_end` case, for example, ...
```
    return parsed != nullptr && !parsed.empty();
```
</blockquote></details>

Done.

Thank you. Made the change here and to the next branch.

**Review by jsrinnn — COMMENTED** [review]
Addressed all feedback. Thank you, Darren!

> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12746#-:-NfRXLhL-oZvnivWTVKm:buht2sm)* status: 9 of 11 files reviewed, 5 unresolved discussions (waiting on @dbolduc and @devbww)

___
*[`google/cloud/bigquery/v2/minimal/benchmarks/job_insert_benchmark_programs.cc` line  at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12746#-NfRL1Mn4cO5M50D130E:-NfRRwUeE6hfwgdmM08e:btwd29y) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/5581524a6beb382b052fb7792275d347c61e3b5d/google/cloud/bigquery/v2/minimal/benchmarks/job_insert_benchmark_programs.cc#L)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

I think we should still leave this out. I did not review it.
</blockquote></details>

Done.

Whoops. That got included with the latest push by mistake, when I was testing the PR with the latest changes. Removed. 

Thank you for pointing it out!!
___
*[`google/cloud/bigquery/v2/minimal/internal/json_utils.cc` line 89 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12746#gh-1339471504:-NfRTVpQ9RVoRF7wAOCX:bb6sh2i) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/f050a835a3c70476b51b990846aa8828a7f16c69/google/cloud/bigquery/v2/minimal/internal/json_utils.cc#L89)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

Thanks, and SGTM.
</blockquote></details>

Ack.
___
*[`google/cloud/bigquery/v2/minimal/internal/json_utils.cc` line 98 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12746#gh-1339472042:-NfRTmrzDPXQBabT22tE:bagmvjy) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/f050a835a3c70476b51b990846aa8828a7f16c69/google/cloud/bigquery/v2/minimal/internal/json_utils.cc#L98)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

The naming is backwards now. As written, if `skip_key` is true, we keep the key. Let me try again with the naming....

I think we should just say:
```cc
        if (event == nlohmann::json::parse_event_t::key) {
          auto const discard = std::any_of(
              keys.begin(), keys.end(),
              [&parsed](std::string const& key) { return parsed == key; });
          return !discard;
        }
```

My naming comment (nit) was about remembering what we are trying to do when a key is found. Someone reading the code will know that `std::any_of` returns true when a match is found.

The only context is the name of the function, which says "RemoveJsonKeys...". We can add a little bit more context in the variable name.
</blockquote></details>

Done.

Correct. if `skip_key` is true then we need to skip the key but the return statement becomes confusing with `return !skip_key`. That is what I was trying to initally point out.

Regardless, I have made the change as I understand your pov as well.
___
*[`google/cloud/bigquery/v2/minimal/internal/json_utils.cc` line 108 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12746#gh-1339491465:-NfRUX6L2s6Y9Otjok1p:bkvk1t4) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/f050a835a3c70476b51b990846aa8828a7f16c69/google/cloud/bigquery/v2/minimal/internal/json_utils.cc#L108)):*
<details><summary><i>Previously, jsrinnn wrote…</i></summary><blockquote>

Done.

Thank you. Made the change here and to the next branch.
</blockquote></details>

Just clarifying the comment above. I meant the next "if" condition with `array_end` and not any code branch. 
___
*[`google/cloud/bigquery/v2/minimal/internal/json_utils_test.cc` line 210 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12746#-NfRQDud8PfLWuSplC84:-NfRWYnc1UV8kWw_oAuG:blweyr7) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/5581524a6beb382b052fb7792275d347c61e3b5d/google/cloud/bigquery/v2/minimal/internal/json_utils_test.cc#L210)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

optional style nit: I think we tend to say `auto constexpr kJsonText = ...` instead of using `auto const*`. I don't actually know the difference.
</blockquote></details>

Done.

made changes here and everwhere else in this file.

**Review by jsrinnn — COMMENTED** [review]
Addressed all feedback. Thank you, Darren!

> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12746#-:-NfRk2KGEHo7Wimrjfaq:buht2sm)* status: 8 of 11 files reviewed, 5 unresolved discussions (waiting on @dbolduc and @devbww)

**Review by devbww — APPROVED** [review]
